File: src/cli/commands/task.ts
import type { Command } from "commander";
import { normalizeStatus, parsePriority } from "../../app/runtime";
import { readStdinContent } from "../../app/stdin";
import { TsqError } from "../../errors";
import type { RunAction, RuntimeDeps } from "../action";
import {
  type CreateCommandOptions,
  type ListCommandOptions,
  type StaleCommandOptions,
  type UpdateCommandOptions,
  applyTreeDefaults,
  asOptionalString,
  collectCsvOption,
  parseKind,
  parseLane,
  parseListFilter,
  parseNonNegativeInt,
  parsePlanningState,
  parsePositiveInt,
  validateExplicitId,
} from "../parsers";
import { printMergeResult, printTask, printTaskList, printTaskTree } from "../render";

export function registerTaskCommands(
  program: Command,
  deps: RuntimeDeps,
  runAction: RunAction,
): void {
  program
    .command("create")
    .argument("<title>", "task title")
    .option("--kind <kind>", "task kind: task|feature|epic", "task")
    .option("-p, --priority <priority>", "priority 0..3", "2")
    .option("--parent <id>", "parent task ID")
    .option("--description <text>", "task description")
    .option("--external-ref <ref>", "external reference (ticket/URL/id)")
    .option("--planning <state>", "planning state: needs_planning|planned")
    .option("--needs-planning", "shorthand for --planning needs_planning")
    .option("--id <id>", "explicit task ID (tsq-<8 crockford base32>)")
    .option("--body-file <path>", "read description from file (use - for stdin)")
    .description("Create task")
    .action(async function action(title: string, options: CreateCommandOptions) {
      await runAction(
        this,
        async (opts) => {
          const kind = parseKind(options.kind ?? "task");
          const priority = parsePriority(options.priority ?? "2");
          if (options.planning && options.needsPlanning) {
            throw new TsqError(
              "VALIDATION_ERROR",
              "cannot combine --planning with --needs-planning",
              1,
            );
          }
          if (options.id && options.parent) {
            throw new TsqError("VALIDATION_ERROR", "cannot combine --id with --parent", 1);
          }
          if (asOptionalString(options.description) !== undefined && options.bodyFile) {
            throw new TsqError(
              "VALIDATION_ERROR",
              "cannot combine --description with --body-file",
              1,
            );
          }
          const planning_state = options.needsPlanning
            ? ("needs_planning" as const)
            : options.planning
              ? parsePlanningState(options.planning)
              : undefined;

          let bodyFile: string | undefined;
          if (options.bodyFile) {
            if (options.bodyFile === "-") {
              bodyFile = await readStdinContent();
            } else {
              const { readFile } = await import("node:fs/promises");
              bodyFile = await readFile(options.bodyFile, "utf8");
            }
            if (bodyFile.trim().length === 0) {
              throw new TsqError("VALIDATION_ERROR", "body file content must not be empty", 1);
            }
          }

          return deps.service.create({
            title,
            kind,
            priority,
            parent: options.parent,
            description: asOptionalString(options.description),
            externalRef: asOptionalString(options.externalRef),
            exactId: opts.exactId,
            planning_state,
            explicitId: options.id ? validateExplicitId(options.id) : undefined,
            bodyFile,
          });
        },
        {
          jsonData: (task) => ({ task }),
          human: (task) => printTask(task),
        },
      );
    });

  program
    .command("show")
    .argument("<id>", "task id")
    .description("Show task details")
    .action(async function action(id: string) {
      await runAction(this, async (opts) => deps.service.show(id, opts.exactId), {
        jsonData: (data) => data,
        human: (data) => {
          printTask(data.task);
          if (data.blockers.length > 0) {
            console.log(`blockers=${data.blockers.join(",")}`);
          }
          if (data.dependents.length > 0) {
            console.log(`dependents=${data.dependents.join(",")}`);
          }
          console.log(`ready=${data.ready}`);
          if (Object.keys(data.links).length > 0) {
            console.log(`links=${JSON.stringify(data.links)}`);
          }
          if (data.history.length > 0) {
            console.log(`history_events=${data.history.length}`);
          }
        },
      });
    });

  const listFlagState = {
    assignee: false,
    unassigned: false,
  };

  program
    .command("list")
    .option("--status <status>", "filter by status")
    .option("--assignee <assignee>", "filter by assignee")
    .option("--external-ref <ref>", "filter by external reference")
    .option("--kind <kind>", "filter by kind")
    .option("--label <label>", "filter by label")
    .option(
      "--label-any <label>",
      "filter by any label (repeatable, comma-separated)",
      collectCsvOption("label-any"),
      [],
    )
    .option("--created-after <iso>", "filter by created_at after timestamp")
    .option("--updated-after <iso>", "filter by updated_at after timestamp")
    .option("--closed-after <iso>", "filter by closed_at after timestamp")
    .option("--unassigned", "filter tasks without assignee")
    .option(
      "--id <id>",
      "filter by task id (repeatable, comma-separated)",
      collectCsvOption("id"),
      [],
    )
    .option("--tree", "render parent/child hierarchy")
    .option("--full", "with --tree, include closed/blocked/canceled items")
    .option("--planning <state>", "filter by planning state: needs_planning|planned")
    .description("List tasks")
    .on("option:assignee", () => {
      listFlagState.assignee = true;
    })
    .on("option:unassigned", () => {
      listFlagState.unassigned = true;
    })
    .action(async function action(options: ListCommandOptions) {
      try {
        const filter = parseListFilter(options, listFlagState.unassigned, listFlagState.assignee);
        if (options.tree) {
          await runAction(
            this,
            async () => deps.service.listTree(applyTreeDefaults(filter, options)),
            {
              jsonData: (tree) => ({ tree }),
              human: (tree) => printTaskTree(tree),
            },
          );
          return;
        }

        await runAction(
          this,
          async () => {
            if (options.full) {
              throw new TsqError("VALIDATION_ERROR", "--full requires --tree", 1);
            }
            return deps.service.list(filter);
          },
          {
            jsonData: (tasks) => ({ tasks }),
            human: (tasks) => printTaskList(tasks),
          },
        );
      } finally {
        listFlagState.assignee = false;
        listFlagState.unassigned = false;
      }
    });

  program
    .command("stale")
    .option("--days <n>", "stale threshold in days", "30")
    .option("--status <status>", "single status scope")
    .option("--assignee <assignee>", "filter by assignee")
    .option("--limit <n>", "max stale tasks to return")
    .description("List stale tasks")
    .action(async function action(options: StaleCommandOptions) {
      await runAction(
        this,
        async () =>
          deps.service.stale({
            days: parseNonNegativeInt(options.days ?? "30", "days"),
            status: options.status ? normalizeStatus(options.status) : undefined,
            assignee: asOptionalString(options.assignee),
            limit: options.limit ? parsePositiveInt(options.limit, "limit", 1, 10000) : undefined,
          }),
        {
          jsonData: (data) => data,
          human: (data) => printTaskList(data.tasks),
        },
      );
    });

  program
    .command("ready")
    .option("--lane <lane>", "filter by lane: planning|coding")
    .description("List ready tasks")
    .action(async function action(options: { lane?: string }) {
      await runAction(
        this,
        async () => {
          const lane = options.lane ? parseLane(options.lane) : undefined;
          return deps.service.ready(lane);
        },
        {
          jsonData: (tasks) => ({ tasks }),
          human: (tasks) => printTaskList(tasks),
        },
      );
    });

  program
    .command("update")
    .argument("<id>", "task id")
    .option("--title <title>", "new title")
    .option("--description <text>", "set task description")
    .option("--clear-description", "clear task description")
    .option("--external-ref <ref>", "set external reference")
    .option("--clear-external-ref", "clear external reference")
    .option("--status <status>", "status value")
    .option("--priority <priority>", "priority 0..3")
    .option("--claim", "claim this task")
    .option("--assignee <assignee>", "assignee for claim")
    .option("--require-spec", "with --claim, require attached spec to pass validation")
    .option("--planning <state>", "set planning state: needs_planning|planned")
    .description("Update task")
    .action(async function action(id: string, options: UpdateCommandOptions) {
      await runAction(
        this,
        async (opts) => {
          const claim = Boolean(options.claim);
          const requireSpec = Boolean(options.requireSpec);
          const hasDescription = asOptionalString(options.description) !== undefined;
          const clearDescription = Boolean(options.clearDescription);
          const hasExternalRef = asOptionalString(options.externalRef) !== undefined;
          const clearExternalRef = Boolean(options.clearExternalRef);
          if (hasDescription && clearDescription) {
            throw new TsqError(
              "VALIDATION_ERROR",
              "cannot combine --description with --clear-description",
              1,
            );
          }
          if (hasExternalRef && clearExternalRef) {
            throw new TsqError(
              "VALIDATION_ERROR",
              "cannot combine --external-ref with --clear-external-ref",
              1,
            );
          }
          if (!claim && requireSpec) {
            throw new TsqError("VALIDATION_ERROR", "--require-spec requires --claim", 1);
          }
          if (claim) {
            if (
              options.title ||
              options.status ||
              options.priority ||
              hasDescription ||
              clearDescription ||
              hasExternalRef ||
              clearExternalRef
            ) {
              throw new TsqError(
                "VALIDATION_ERROR",
                "cannot combine --claim with --title/--description/--clear-description/--external-ref/--clear-external-ref/--status/--priority",
                1,
              );
            }
            return deps.service.claim({
              id,
              assignee: asOptionalString(options.assignee),
              requireSpec,
              exactId: opts.exactId,
            });
          }
          return deps.service.update({
            id,
            title: asOptionalString(options.title),
            description: asOptionalString(options.description),
            clearDescription,
            externalRef: asOptionalString(options.externalRef),
            clearExternalRef,
            status: options.status ? normalizeStatus(String(options.status)) : undefined,
            priority: options.priority ? parsePriority(String(options.priority)) : undefined,
            planning_state: options.planning ? parsePlanningState(options.planning) : undefined,
            exactId: opts.exactId,
          });
        },
        {
          jsonData: (task) => ({ task }),
          human: (task) => printTask(task),
        },
      );
    });

  program
    .command("duplicate")
    .argument("<id>", "duplicate task id")
    .requiredOption("--of <canonical-id>", "canonical task id")
    .option("--reason <text>", "duplicate reason")
    .description("Mark a task as duplicate of canonical task")
    .action(async function action(id: string, options: { of: string; reason?: string }) {
      await runAction(
        this,
        async (opts) =>
          deps.service.duplicate({
            source: id,
            canonical: options.of,
            reason: options.reason,
            exactId: opts.exactId,
          }),
        {
          jsonData: (task) => ({ task }),
          human: (task) => printTask(task),
        },
      );
    });

  program
    .command("duplicates")
    .option("--limit <n>", "max duplicate groups", "20")
    .description("Dry-run duplicate detector scaffold")
    .action(async function action(options: { limit?: string }) {
      await runAction(
        this,
        async () =>
          deps.service.duplicateCandidates(
            parsePositiveInt(options.limit ?? "20", "limit", 1, 200),
          ),
        {
          jsonData: (data) => data,
          human: (data) => {
            if (data.groups.length === 0) {
              console.log("no duplicate candidates");
              return;
            }
            console.log(`scanned=${data.scanned} groups=${data.groups.length}`);
            for (const group of data.groups) {
              const ids = group.tasks.map((task) => task.id).join(",");
              console.log(`${group.key}: ${ids}`);
            }
          },
        },
      );
    });

  program
    .command("supersede")
    .argument("<old-id>", "source task id")
    .requiredOption("--with <new-id>", "replacement task id")
    .option("--reason <text>", "supersede reason")
    .description("Supersede a task with another")
    .action(async function action(oldId: string, options: { with: string; reason?: string }) {
      await runAction(
        this,
        async (opts) =>
          deps.service.supersede({
            source: oldId,
            withId: options.with,
            reason: options.reason,
            exactId: opts.exactId,
          }),
        {
          jsonData: (task) => ({ task }),
          human: (task) => printTask(task),
        },
      );
    });

  program
    .command("merge")
    .argument("<sources...>", "source task IDs to merge")
    .requiredOption("--into <target>", "target task ID to merge into")
    .option("--reason <text>", "merge reason")
    .option("--force", "allow merging into closed/canceled target")
    .option("--dry-run", "preview merge without applying")
    .description("Merge tasks as duplicates into a target")
    .action(async function action(
      sources: string[],
      options: {
        into: string;
        reason?: string;
        force?: boolean;
        dryRun?: boolean;
      },
    ) {
      await runAction(
        this,
        async (opts) =>
          deps.service.merge({
            sources,
            into: options.into,
            reason: options.reason,
            force: Boolean(options.force),
            dryRun: Boolean(options.dryRun),
            exactId: opts.exactId,
          }),
        {
          jsonData: (data) => data,
          human: (data) => printMergeResult(data),
        },
      );
    });

  program
    .command("close")
    .argument("<ids...>", "task ids to close")
    .option("--reason <text>", "close reason")
    .description("Close tasks")
    .action(async function action(ids: string[], options: { reason?: string }) {
      await runAction(
        this,
        async (opts) => deps.service.close({ ids, reason: options.reason, exactId: opts.exactId }),
        {
          jsonData: (tasks) => ({ tasks }),
          human: (tasks) => {
            for (const task of tasks) {
              printTask(task);
            }
          },
        },
      );
    });

  program
    .command("reopen")
    .argument("<ids...>", "task ids to reopen")
    .description("Reopen closed tasks")
    .action(async function action(ids: string[]) {
      await runAction(this, async (opts) => deps.service.reopen({ ids, exactId: opts.exactId }), {
        jsonData: (tasks) => ({ tasks }),
        human: (tasks) => {
          for (const task of tasks) {
            printTask(task);
          }
        },
      });
    });

  program
    .command("search")
    .argument("<query>", "search query")
    .description("Search tasks")
    .action(async function action(query: string) {
      await runAction(this, async () => deps.service.search({ query }), {
        jsonData: (tasks) => ({ tasks }),
        human: (tasks) => printTaskList(tasks),
      });
    });
}

File: src/cli/commands/task.rs
use crate::app::runtime::normalize_status;
use crate::app::service::TasqueService;
use crate::app::service_query::ShowResult;
use crate::app::service_types::{
    ClaimInput, DuplicateInput, MergeInput, SpecContentInput, SpecContentResult, StaleInput,
    SupersedeInput, UpdateInput,
};
use crate::cli::action::{GlobalOpts, run_action};
use crate::cli::parsers::{
    as_optional_string, parse_non_negative_int, parse_positive_int, parse_priority_value,
};
use crate::cli::render::{
    print_history, print_history_plain, print_merge_result, print_show_result,
    print_show_result_plain, print_spec_content, print_spec_content_plain, print_task,
    print_task_list, print_task_notes, print_task_notes_plain,
};
use crate::errors::TsqError;
use clap::Args;

#[path = "task_create.rs"]
mod task_create;
#[path = "task_find.rs"]
mod task_find;
#[path = "task_lifecycle.rs"]
mod task_lifecycle;

pub use task_create::{CreateArgs, execute_create};
pub use task_find::{FindArgs, execute_find};
pub use task_lifecycle::{
    MultiStatusArgs, NoteStatusArgs, execute_cancel, execute_defer, execute_done, execute_reopen,
};

#[derive(Debug, Args)]
pub struct ShowArgs {
    pub id: String,
    #[arg(long = "with-spec", alias = "spec", default_value_t = false)]
    pub with_spec: bool,
    #[arg(long, default_value_t = false)]
    pub deps: bool,
    #[arg(long, default_value_t = false)]
    pub notes: bool,
    #[arg(long, default_value_t = false)]
    pub history: bool,
    #[arg(long = "all-notes", default_value_t = false)]
    pub all_notes: bool,
    #[arg(long = "all-history", default_value_t = false)]
    pub all_history: bool,
}

#[derive(Debug, Args)]
pub struct StaleArgs {
    #[arg(long, default_value = "30")]
    pub days: String,
    #[arg(long)]
    pub status: Option<String>,
    #[arg(long)]
    pub assignee: Option<String>,
    #[arg(long)]
    pub limit: Option<String>,
}

#[derive(Debug, Args)]
pub struct EditArgs {
    pub id: String,
    #[arg(long)]
    pub title: Option<String>,
    #[arg(long)]
    pub description: Option<String>,
    #[arg(long = "clear-description", default_value_t = false)]
    pub clear_description: bool,
    #[arg(long = "external-ref")]
    pub external_ref: Option<String>,
    #[arg(long = "discovered-from")]
    pub discovered_from: Option<String>,
    #[arg(long = "clear-discovered-from", default_value_t = false)]
    pub clear_discovered_from: bool,
    #[arg(long = "clear-external-ref", default_value_t = false)]
    pub clear_external_ref: bool,
    #[arg(long)]
    pub priority: Option<String>,
}

#[derive(Debug, Args)]
pub struct ClaimArgs {
    pub id: String,
    #[arg(long)]
    pub assignee: Option<String>,
    #[arg(long, default_value_t = false)]
    pub start: bool,
    #[arg(long = "require-spec", default_value_t = false)]
    pub require_spec: bool,
}

#[derive(Debug, Args)]
pub struct WorkonArgs {
    pub id: String,
    #[arg(long)]
    pub assignee: Option<String>,
    #[arg(long = "require-spec", default_value_t = false)]
    pub require_spec: bool,
}

#[derive(Debug, Args)]
pub struct AssignArgs {
    pub id: String,
    #[arg(long)]
    pub assignee: String,
}

#[derive(Debug, Args)]
pub struct TaskIdArgs {
    pub id: String,
}

#[derive(Debug, Args)]
pub struct DuplicateArgs {
    pub id: String,
    pub of: String,
    pub canonical: String,
    #[arg(long)]
    pub note: Option<String>,
}

#[derive(Debug, Args)]
pub struct DuplicatesArgs {
    #[arg(long, default_value = "20")]
    pub limit: String,
}

#[derive(Debug, Args)]
pub struct SupersedeArgs {
    pub old_id: String,
    pub with: String,
    pub new_id: String,
    #[arg(long)]
    pub note: Option<String>,
}

#[derive(Debug, Args)]
pub struct MergeArgs {
    pub sources: Vec<String>,
    #[arg(long = "into")]
    pub into: String,
    #[arg(long)]
    pub reason: Option<String>,
    #[arg(long, default_value_t = false)]
    pub force: bool,
    #[arg(long = "dry-run", default_value_t = false)]
    pub dry_run: bool,
}

pub fn execute_show(service: &TasqueService, args: ShowArgs, opts: GlobalOpts) -> i32 {
    let selected = args.with_spec || args.deps || args.notes || args.history;
    run_action(
        "tsq show",
        opts,
        || {
            let show = service.show(&args.id, opts.exact_id)?;
            let spec_input = SpecContentInput {
                id: args.id.clone(),
                exact_id: opts.exact_id,
            };
            let spec = if args.with_spec {
                Some(service.spec_content(spec_input)).transpose()?
            } else if !selected {
                service.spec_content(spec_input).ok()
            } else {
                None
            };
            Ok((show, spec))
        },
        |(show, spec)| {
            let notes_limit = if args.all_notes { usize::MAX } else { 5 };
            let history_limit = if args.all_history { usize::MAX } else { 10 };
            show_data_json(
                service,
                show,
                spec.as_ref(),
                !selected || args.deps,
                !selected || args.with_spec,
                !selected || args.notes,
                !selected || args.history,
                notes_limit,
                history_limit,
            )
        },
        |(show, spec)| {
            let notes_limit = if args.all_notes { usize::MAX } else { 5 };
            let history_limit = if args.all_history { usize::MAX } else { 10 };
            if opts.plain() {
                if !selected || args.deps {
                    print_show_result_plain(show);
                }
                if let Some(spec) = spec {
                    print_spec_content_plain(spec);
                }
                if !selected || args.notes {
                    let notes = show
                        .task
                        .notes
                        .iter()
                        .take(notes_limit)
                        .cloned()
                        .collect::<Vec<_>>();
                    print_task_notes_plain(&show.task.id, &notes);
                }
                if !selected || args.history {
                    let events = show
                        .history
                        .iter()
                        .rev()
                        .take(history_limit)
                        .cloned()
                        .collect::<Vec<_>>();
                    print_history_plain(&crate::app::service_types::HistoryResult {
                        count: events.len(),
                        truncated: show.history.len() > events.len(),
                        events,
                    });
                }
                return Ok(());
            }
            if !selected || args.deps {
                print_show_result(show);
            }
            if let Some(spec) = spec {
                print_spec_content(spec);
            }
            if !selected || args.notes {
                let notes = show
                    .task
                    .notes
                    .iter()
                    .take(notes_limit)
                    .cloned()
                    .collect::<Vec<_>>();
                print_task_notes(&show.task.id, &notes);
            }
            if !selected || args.history {
                let events = show
                    .history
                    .iter()
                    .rev()
                    .take(history_limit)
                    .cloned()
                    .collect::<Vec<_>>();
                print_history(&crate::app::service_types::HistoryResult {
                    count: events.len(),
                    truncated: show.history.len() > events.len(),
                    events,
                });
            }
            Ok(())
        },
    )
}

#[allow(clippy::too_many_arguments)]
fn show_data_json(
    service: &TasqueService,
    show: &ShowResult,
    spec: Option<&SpecContentResult>,
    include_deps: bool,
    include_spec: bool,
    include_notes: bool,
    include_history: bool,
    notes_limit: usize,
    history_limit: usize,
) -> serde_json::Value {
    let mut json = serde_json::json!({
        "task": show.task,
        "root": service.repo_root(),
    });
    let data = json.as_object_mut().expect("show JSON object");
    if include_deps {
        data.insert(
            "deps".to_string(),
            serde_json::json!({
                "blockers": show.blocker_edges,
                "dependents": show.dependent_edges,
                "ready": show.ready,
                "links": show.links,
            }),
        );
    }
    if include_spec {
        data.insert("spec".to_string(), serde_json::json!(spec));
    }
    if include_notes {
        data.insert(
            "notes".to_string(),
            serde_json::json!(show.task.notes.iter().take(notes_limit).collect::<Vec<_>>()),
        );
    }
    if include_history {
        data.insert(
            "history".to_string(),
            serde_json::json!(
                show.history
                    .iter()
                    .rev()
                    .take(history_limit)
                    .collect::<Vec<_>>()
            ),
        );
    }
    json
}

pub fn execute_stale(service: &TasqueService, args: StaleArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq stale",
        opts,
        || {
            let days = parse_non_negative_int(&args.days, "days")?;
            let status = args.status.as_deref().map(normalize_status).transpose()?;
            let limit = args
                .limit
                .as_deref()
                .map(|value| parse_positive_int(value, "limit", 1, 10000))
                .transpose()?
                .map(|value| value as usize);
            service.stale(&StaleInput {
                days,
                status,
                assignee: as_optional_string(args.assignee.as_deref()),
                limit,
            })
        },
        |data| data.clone(),
        |data| {
            print_task_list(&data.tasks);
            Ok(())
        },
    )
}

pub fn execute_edit(service: &TasqueService, args: EditArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq edit",
        opts,
        || {
            let has_description = as_optional_string(args.description.as_deref()).is_some();
            let clear_description = args.clear_description;
            let has_external_ref = as_optional_string(args.external_ref.as_deref()).is_some();
            let has_discovered_from = as_optional_string(args.discovered_from.as_deref()).is_some();
            let clear_external_ref = args.clear_external_ref;
            let clear_discovered_from = args.clear_discovered_from;

            if has_description && clear_description {
                return Err(TsqError::new(
                    "VALIDATION_ERROR",
                    "cannot combine --description with --clear-description",
                    1,
                ));
            }
            if has_external_ref && clear_external_ref {
                return Err(TsqError::new(
                    "VALIDATION_ERROR",
                    "cannot combine --external-ref with --clear-external-ref",
                    1,
                ));
            }
            if has_discovered_from && clear_discovered_from {
                return Err(TsqError::new(
                    "VALIDATION_ERROR",
                    "cannot combine --discovered-from with --clear-discovered-from",
                    1,
                ));
            }
            service.update(UpdateInput {
                id: args.id.clone(),
                title: as_optional_string(args.title.as_deref()),
                description: as_optional_string(args.description.as_deref()),
                clear_description,
                external_ref: as_optional_string(args.external_ref.as_deref()),
                discovered_from: as_optional_string(args.discovered_from.as_deref()),
                clear_discovered_from,
                clear_external_ref,
                status: None,
                priority: args
                    .priority
                    .as_deref()
                    .map(parse_priority_value)
                    .transpose()?,
                exact_id: opts.exact_id,
                planning_state: None,
                assignee: None,
            })
        },
        |task| serde_json::json!({ "task": task }),
        |task| {
            print_task(task);
            Ok(())
        },
    )
}

pub fn execute_duplicate(service: &TasqueService, args: DuplicateArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq duplicate",
        opts,
        || {
            validate_sentence_token(&args.of, "of", "tsq duplicate <id> of <canonical>")?;
            service.duplicate(DuplicateInput {
                source: args.id.clone(),
                canonical: args.canonical.clone(),
                reason: args.note.clone(),
                exact_id: opts.exact_id,
            })
        },
        |task| serde_json::json!({ "task": task }),
        |task| {
            print_task(task);
            Ok(())
        },
    )
}

pub fn execute_duplicates(service: &TasqueService, args: DuplicatesArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq duplicates",
        opts,
        || {
            let limit = parse_positive_int(&args.limit, "limit", 1, 200)? as usize;
            service.duplicate_candidates(Some(limit))
        },
        |data| data.clone(),
        |data| {
            if data.groups.is_empty() {
                println!("no duplicate candidates");
                return Ok(());
            }
            println!("scanned={} groups={}", data.scanned, data.groups.len());
            for group in &data.groups {
                let ids = group
                    .tasks
                    .iter()
                    .map(|task| task.id.as_str())
                    .collect::<Vec<_>>()
                    .join(",");
                println!("{}: {}", group.key, ids);
            }
            Ok(())
        },
    )
}

pub fn execute_supersede(service: &TasqueService, args: SupersedeArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq supersede",
        opts,
        || {
            validate_sentence_token(&args.with, "with", "tsq supersede <old> with <new>")?;
            service.supersede(SupersedeInput {
                source: args.old_id.clone(),
                with_id: args.new_id.clone(),
                reason: args.note.clone(),
                exact_id: opts.exact_id,
            })
        },
        |task| serde_json::json!({ "task": task }),
        |task| {
            print_task(task);
            Ok(())
        },
    )
}

pub fn execute_merge(service: &TasqueService, args: MergeArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq merge",
        opts,
        || {
            service.merge(MergeInput {
                sources: args.sources.clone(),
                into: args.into.clone(),
                reason: args.reason.clone(),
                force: args.force,
                dry_run: args.dry_run,
                exact_id: opts.exact_id,
            })
        },
        |data| data.clone(),
        |data| {
            print_merge_result(data);
            Ok(())
        },
    )
}

pub fn execute_claim(service: &TasqueService, args: ClaimArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq claim",
        opts,
        || {
            let _ = args.start;
            service.claim(ClaimInput {
                id: args.id.clone(),
                assignee: as_optional_string(args.assignee.as_deref()),
                start: args.start,
                require_spec: args.require_spec,
                exact_id: opts.exact_id,
            })
        },
        |task| serde_json::json!({ "task": task }),
        |task| {
            print_task(task);
            Ok(())
        },
    )
}

pub fn execute_workon(service: &TasqueService, args: WorkonArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq workon",
        opts,
        || {
            service.claim(ClaimInput {
                id: args.id.clone(),
                assignee: as_optional_string(args.assignee.as_deref()),
                start: true,
                require_spec: args.require_spec,
                exact_id: opts.exact_id,
            })?;
            let show = service.show(&args.id, opts.exact_id)?;
            let spec_input = SpecContentInput {
                id: args.id.clone(),
                exact_id: opts.exact_id,
            };
            let spec = service.spec_content(spec_input).ok();
            Ok((show, spec))
        },
        |(show, spec)| show_data_json(service, show, spec.as_ref(), true, true, true, true, 5, 10),
        |(show, _spec)| {
            if opts.plain() {
                crate::cli::render::print_show_result_plain(show);
            } else {
                print_show_result(show);
            }
            Ok(())
        },
    )
}

pub fn execute_assign(service: &TasqueService, args: AssignArgs, opts: GlobalOpts) -> i32 {
    run_action(
        "tsq assign",
        opts,
        || {
            service.update(UpdateInput {
                id: args.id.clone(),
                title: None,
                description: None,
                clear_description: false,
                external_ref: None,
                discovered_from: None,
                clear_discovered_from: false,
                clear_external_ref: false,
                status: None,
                priority: None,
                exact_id: opts.exact_id,
                planning_state: None,
                assignee: as_optional_string(Some(&args.assignee)),
            })
        },
        |task| serde_json::json!({ "task": task }),
        |task| {
            print_task(task);
            Ok(())
        },
    )
}

pub fn execute_set_status(
    service: &TasqueService,
    args: TaskIdArgs,
    status: crate::types::TaskStatus,
    command_line: &'static str,
    opts: GlobalOpts,
) -> i32 {
    run_action(
        command_line,
        opts,
        || {
            service.update(UpdateInput {
                id: args.id.clone(),
                title: None,
                description: None,
                clear_description: false,
                external_ref: None,
                discovered_from: None,
                clear_discovered_from: false,
                clear_external_ref: false,
                status: Some(status),
                priority: None,
                exact_id: opts.exact_id,
                planning_state: None,
                assignee: None,
            })
        },
        |task| serde_json::json!({ "task": task }),
        |task| {
            print_task(task);
            Ok(())
        },
    )
}

pub fn execute_set_planning(
    service: &TasqueService,
    args: TaskIdArgs,
    planning_state: crate::types::PlanningState,
    command_line: &'static str,
    opts: GlobalOpts,
) -> i32 {
    run_action(
        command_line,
        opts,
        || {
            service.update(UpdateInput {
                id: args.id.clone(),
                title: None,
                description: None,
                clear_description: false,
                external_ref: None,
                discovered_from: None,
                clear_discovered_from: false,
                clear_external_ref: false,
                status: None,
                priority: None,
                exact_id: opts.exact_id,
                planning_state: Some(planning_state),
                assignee: None,
            })
        },
        |task| serde_json::json!({ "task": task }),
        |task| {
            print_task(task);
            Ok(())
        },
    )
}

fn validate_sentence_token(value: &str, expected: &str, example: &str) -> Result<(), TsqError> {
    if value == expected {
        return Ok(());
    }
    Err(TsqError::new(
        "VALIDATION_ERROR",
        format!("expected `{}`; use `{}`", expected, example),
        1,
    ))
}
